File: ios/wrappers/inapp/ReactNotificationInboxChangeListener.swift
import CioMessagingInApp
import Foundation

/// React Native wrapper for NotificationInboxChangeListener.
/// Singleton instance that forwards inbox message changes to React Native event emitter.
/// All methods run on MainActor as required by the protocol.
@MainActor
class ReactNotificationInboxChangeListener: NotificationInboxChangeListener {
    // Event emitter function to send events to React Native layer
    private var eventEmitter: (([String: Any]) -> Void)?

    // Sets the event emitter function
    func setEventEmitter(emitter: (([String: Any]) -> Void)?) {
        eventEmitter = emitter
    }

    // Clears the event emitter to prevent memory leaks
    func clearEventEmitter() {
        eventEmitter = nil
    }

    // Called when inbox messages change
    func onMessagesChanged(messages: [InboxMessage]) {
        guard let emitter = eventEmitter else {
            return
        }

        let messagesArray = messages.map { $0.toDictionary() }
        let payload: [String: Any] = ["messages": messagesArray]

        emitter(payload)
    }

    // Singleton instance
    static let shared = ReactNotificationInboxChangeListener()
    private init() {}
}

File: android/src/main/java/io/customer/reactnative/sdk/messaginginapp/ReactNotificationInboxChangeListener.kt
package io.customer.reactnative.sdk.messaginginapp

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReadableMap
import io.customer.messaginginapp.gist.data.model.InboxMessage
import io.customer.messaginginapp.inbox.NotificationInboxChangeListener

class ReactNotificationInboxChangeListener private constructor() : NotificationInboxChangeListener {

    // Event emitter function to send events to React Native layer
    private var eventEmitter: ((ReadableMap) -> Unit)? = null

    // Sets the event emitter function and message converter
    internal fun setEventEmitter(emitter: ((ReadableMap) -> Unit)?) {
        this.eventEmitter = emitter
    }

    // Clears the event emitter to prevent memory leaks
    internal fun clearEventEmitter() {
        this.eventEmitter = null
    }

    private fun emitMessagesUpdate(messages: List<InboxMessage>) {
        // Get the emitter and converter, return early if not set
        val emitter = eventEmitter ?: return

        val messagesArray = Arguments.createArray()
        messages.forEach { message ->
            messagesArray.pushMap(message.toWritableMap())
        }

        val payload = Arguments.createMap().apply {
            putArray("messages", messagesArray)
        }

        emitter.invoke(payload)
    }

    override fun onMessagesChanged(messages: List<InboxMessage>) {
        emitMessagesUpdate(messages)
    }

    companion object {
        // Singleton instance with public visibility for direct access
        val instance: ReactNotificationInboxChangeListener by lazy { ReactNotificationInboxChangeListener() }
    }
}
